From 890cfc08bf4ac870bd55c1afa4030ff69adea770 Mon Sep 17 00:00:00 2001
From: Feng Chen <feng.chen@amlogic.com>
Date: Tue, 21 Mar 2023 10:59:15 +0800
Subject: [PATCH] mtdblock: modify for squashfs mount [1/2]

PD#SWPL-116004

Problem:
1.squashfs don't handle bad block
2.partition size returned from mtdblock
is incorrect

Solution:
1.mtdblock handle bad block
2.fix mtdblock size

Verify:
s4_aq2432

Change-Id: Icbaec2b6679727ebe0d12fd92bf4d21cc4d5f3fe
Signed-off-by: Feng Chen <feng.chen@amlogic.com>
---
 drivers/mtd/mtdblock.c | 107 ++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 106 insertions(+), 1 deletion(-)

diff --git a/drivers/mtd/mtdblock.c b/drivers/mtd/mtdblock.c
index a0a1194dc1d9..1c57b9eccad7 100644
--- a/drivers/mtd/mtdblock.c
+++ b/drivers/mtd/mtdblock.c
@@ -31,6 +31,14 @@ struct mtdblk_dev {
 	enum { STATE_EMPTY, STATE_CLEAN, STATE_DIRTY } cache_state;
 };
 
+#ifdef CONFIG_AMLOGIC_NAND
+struct mtdblk_pbbt {
+	struct mtdblk_dev mb;
+	unsigned int bad_cnt;
+	unsigned short *part_bbt;
+};
+#endif
+
 /*
  * Cache stuff...
  *
@@ -101,11 +109,41 @@ static int write_cached_data (struct mtdblk_dev *mtdblk)
 	 * written to the device. Clear cache_state to avoid writing to
 	 * bad blocks repeatedly.
 	 */
+#ifdef CONFIG_AMLOGIC_NAND
+	if (!ret && ret != EIO)
+		return ret;
+	mtdblk->cache_state = STATE_EMPTY;
+	return ret;
+#else
 	if (ret == 0 || ret == -EIO)
 		mtdblk->cache_state = STATE_EMPTY;
 	return ret;
+#endif
 }
 
+#ifdef CONFIG_AMLOGIC_NAND
+static unsigned long map_block(struct mtdblk_dev *mtdblk, unsigned long pos)
+{
+	struct mtd_info *mtd = mtdblk->mbd.mtd;
+	struct mtdblk_pbbt *pbbt = container_of(mtdblk, struct mtdblk_pbbt, mb);
+	int block, i;
+
+	if (!pbbt->part_bbt)
+		return pos;
+
+	block = (int)(pos >> mtd->erasesize_shift);
+	for (i = 0; i < pbbt->bad_cnt; i++) {
+		if (block >= pbbt->part_bbt[i])
+			block++;
+		else
+			break;
+	}
+
+	/* form actual position */
+	return ((unsigned long)block * mtd->erasesize) |
+		(pos & (mtd->erasesize - 1));
+}
+#endif
 
 static int do_cached_write (struct mtdblk_dev *mtdblk, unsigned long pos,
 			    int len, const char *buf)
@@ -118,6 +156,10 @@ static int do_cached_write (struct mtdblk_dev *mtdblk, unsigned long pos,
 	pr_debug("mtdblock: write on \"%s\" at 0x%lx, size 0x%x\n",
 		mtd->name, pos, len);
 
+#ifdef CONFIG_AMLOGIC_NAND
+	pos = map_block(mtdblk, pos);
+#endif
+
 	if (!sect_size)
 		return mtd_write(mtd, pos, len, &retlen, buf);
 
@@ -187,7 +229,9 @@ static int do_cached_read (struct mtdblk_dev *mtdblk, unsigned long pos,
 
 	pr_debug("mtdblock: read on \"%s\" at 0x%lx, size 0x%x\n",
 			mtd->name, pos, len);
-
+#ifdef CONFIG_AMLOGIC_NAND
+	pos = map_block(mtdblk, pos);
+#endif
 	if (!sect_size) {
 		ret = mtd_read(mtd, pos, len, &retlen, buf);
 		if (ret && !mtd_is_bitflip(ret))
@@ -253,6 +297,10 @@ static int mtdblock_writesect(struct mtd_blktrans_dev *dev,
 static int mtdblock_open(struct mtd_blktrans_dev *mbd)
 {
 	struct mtdblk_dev *mtdblk = container_of(mbd, struct mtdblk_dev, mbd);
+#ifdef CONFIG_AMLOGIC_NAND
+	int block_cnt, i, bad_cnt = 0;
+	struct mtdblk_pbbt *pbbt = container_of(mtdblk, struct mtdblk_pbbt, mb);
+#endif
 
 	pr_debug("mtdblock_open\n");
 
@@ -274,6 +322,35 @@ static int mtdblock_open(struct mtd_blktrans_dev *mbd)
 		mtdblk->cache_data = NULL;
 	}
 
+#ifdef CONFIG_AMLOGIC_NAND
+	pbbt->part_bbt =  NULL;
+	if (!mtd_can_have_bb(mbd->mtd))
+		goto _ok;
+
+	block_cnt = mbd->mtd->size >> mbd->mtd->erasesize_shift;
+	for (i = 0; i < block_cnt; i++)
+		/*
+		 * A valid judgment is made before calling this function.
+		 */
+		/* coverity[divide_by_zero:SUPPRESS] */
+		if (mtd_block_isbad(mbd->mtd, i * mbd->mtd->erasesize))
+			bad_cnt++;
+	pbbt->bad_cnt = bad_cnt;
+	if (bad_cnt) {
+		pbbt->part_bbt =
+		kmalloc_array(block_cnt, sizeof(*pbbt->part_bbt), GFP_KERNEL);
+		bad_cnt = 0;
+		for (i = 0; i < block_cnt; i++)
+			/*
+			 * A valid judgment is made before calling this function.
+			 */
+			/* coverity[divide_by_zero:SUPPRESS] */
+			if (mtd_block_isbad(mbd->mtd, i * mbd->mtd->erasesize))
+				pbbt->part_bbt[bad_cnt++] = i;
+	}
+
+_ok:
+#endif
 	pr_debug("ok\n");
 
 	return 0;
@@ -311,27 +388,55 @@ static int mtdblock_flush(struct mtd_blktrans_dev *dev)
 	ret = write_cached_data(mtdblk);
 	mutex_unlock(&mtdblk->cache_mutex);
 	mtd_sync(dev->mtd);
+
 	return ret;
 }
 
 static void mtdblock_add_mtd(struct mtd_blktrans_ops *tr, struct mtd_info *mtd)
 {
 	struct mtdblk_dev *dev = kzalloc(sizeof(*dev), GFP_KERNEL);
+#ifdef CONFIG_AMLOGIC_NAND
+	int i = 0;
+	struct mtdblk_pbbt *pbbt = kzalloc(sizeof(*pbbt), GFP_KERNEL);
+#endif
 
 	if (!dev)
 		return;
 
+#ifdef CONFIG_AMLOGIC_NAND
+	if (!pbbt)
+		return;
+#endif
+
 	dev->mbd.mtd = mtd;
 	dev->mbd.devnum = mtd->index;
 
 	dev->mbd.size = mtd->size >> 9;
+
+#ifdef CONFIG_AMLOGIC_NAND
+	if (!mtd_can_have_bb(mtd))
+		goto _ok;
+
+	for (i = 0; i < (mtd->size >> mtd->erasesize_shift); i++)
+		if (mtd_block_isbad(mtd, i * mtd->erasesize))
+			dev->mbd.size -= (mtd->erasesize >> 9);
+_ok:
+#endif
 	dev->mbd.tr = tr;
 
 	if (!(mtd->flags & MTD_WRITEABLE))
 		dev->mbd.readonly = 1;
 
+#ifdef CONFIG_AMLOGIC_NAND
+	pbbt->mb = *dev;
+	if (add_mtd_blktrans_dev(&dev->mbd)) {
+		kfree(dev);
+		kfree(pbbt);
+	}
+#else
 	if (add_mtd_blktrans_dev(&dev->mbd))
 		kfree(dev);
+#endif
 }
 
 static void mtdblock_remove_dev(struct mtd_blktrans_dev *dev)
-- 
2.25.1

